
The `uc.pipeline.anchor` function is responsible for analysis of read capture efficiency using anchor-interposed beads. The anchor oligonucleotide design was proposed by [this research](https://www.nature.com/articles/s42003-025-07474-5) :fontawesome-solid-arrow-up-right-from-square:. In mcverse, we applied Tresor to generate a bunch of reads with which we used the anchor pipeline to calculate the percentage of caaptured reads. 

!!! tip "hands-on analysis"

    Please see [details](https://2003100127.github.io/mcverse/bead-anchor) :fontawesome-solid-arrow-up-right-from-square: about a hands-on analysis using the pipeline.

This section demonstrates only the essential components and minimal runnable example. We first define the required parameters in a `.yml` file, which are fully parsed by the relevant modules within UMIche.

It is important to note that `work_dir` should point to the directory where simulated reads have already been generated by [Tresor](https://2003100127.github.io/tresor) :fontawesome-solid-arrow-up-right-from-square: under a specific error scenario.

We need to define a series of varying `criteria` under the `varied` attribute. Typically, we run 10 permutation tests to reduce the impact of random noise. As this is a user-defined sequencing experiment, in principle, you can specify an unlimited number of `seq` values under the `anchor` attribute.

``` yaml
work_dir: ../../data/tresor/anchor/simu/pcr_del/

fixed:
  permutation_num: 10

varied:
    criteria: [
        1e-05,
        2.5e-05,
        5e-05,
        7.5e-05,
        0.0001,
        0.00025,
        0.0005,
        0.00075,
        0.001,
        0.0025,
        0.005,
        0.0075,
        0.01,
        0.025,
        0.05,
        0.075,
        0.1,
        # 0.2,
        # 0.3,
    ]

anchor:
    seq1: 'BAGC'
```

Then, please specify the scenario in the function, again. 

``` py
pct_reads, pct_anchor_reads = uc.pipeline.anchor(
    scenario='pcr_del',
    param_fpn='../../data/tresor/anchor/params_anchor.yml',
)
```

It will calculate % of reads captured without an anchor.

``` shell
{0: [0.9952, 0.9962, 0.9954, 0.9944, 0.9894, 0.9852, 0.9798, 0.9706, 0.9392, 0.8288, 0.771, 0.6808, 0.6126, 0.3122, 0.142, 0.0572, 0.0282], 1: [0.998, 0.9892, 0.9956, 0.9734, 0.978, 0.987, 0.953, 0.9536, 0.9346, 0.9046, 0.8136, 0.7044, 0.5678, 0.3194, 0.094, 0.0358, 0.0324], 2: [0.9972, 0.9956, 0.9956, 0.9948, 0.9702, 0.9614, 0.9672, 0.9506, 0.9042, 0.851, 0.7844, 0.6466, 0.5246, 0.2814, 0.0998, 0.061, 0.0232], 3: [0.9868, 0.9954, 0.994, 0.9956, 0.995, 0.9866, 0.936, 0.9348, 0.95, 0.8762, 0.7068, 0.6174, 0.6248, 0.27, 0.1254, 0.0608, 0.0286], 4: [0.9974, 0.9942, 0.9956, 0.994, 0.988, 0.9632, 0.9546, 0.9348, 0.9498, 0.8892, 0.7662, 0.666, 0.5966, 0.2898, 0.1236, 0.059, 0.0212], 5: [0.9964, 0.996, 0.9922, 0.9936, 0.9914, 0.9666, 0.975, 0.9622, 0.9758, 0.84, 0.7312, 0.6896, 0.5448, 0.2692, 0.0928, 0.0482, 0.0236], 6: [0.996, 0.994, 0.9926, 0.9812, 0.993, 0.9804, 0.9444, 0.9674, 0.9678, 0.8592, 0.782, 0.6528, 0.6428, 0.282, 0.118, 0.0584, 0.0386], 7: [0.9966, 0.9962, 0.9958, 0.9932, 0.9914, 0.9822, 0.9346, 0.9442, 0.9306, 0.8476, 0.8036, 0.697, 0.6026, 0.3232, 0.1032, 0.0376, 0.0272], 8: [0.997, 0.995, 0.9928, 0.9938, 0.9922, 0.988, 0.984, 0.9636, 0.9426, 0.9036, 0.7962, 0.6374, 0.5552, 0.2944, 0.0866, 0.0422, 0.0256], 9: [0.9962, 0.991, 0.9956, 0.9928, 0.9928, 0.9856, 0.9802, 0.9528, 0.9366, 0.9044, 0.7412, 0.7106, 0.5484, 0.3078, 0.1202, 0.0418, 0.0264]}
```

Also, it will calculate % of reads captured with an anchor.

``` shell
{0: [0.981, 0.9402, 0.9622, 0.8878, 0.8486, 0.6536, 0.4906, 0.3792, 0.287, 0.0862, 0.0192, 0.003, 0.0008, 0.0, 0.0, 0.0, 0.0], 1: [0.9938, 0.9716, 0.9328, 0.8968, 0.868, 0.7362, 0.5006, 0.393, 0.2776, 0.0768, 0.0072, 0.003, 0.0018, 0.0, 0.0, 0.0, 0.0], 2: [0.994, 0.974, 0.917, 0.9164, 0.8914, 0.6954, 0.5458, 0.3634, 0.297, 0.0644, 0.009, 0.0026, 0.001, 0.0, 0.0, 0.0, 0.0], 3: [0.9766, 0.969, 0.9596, 0.8992, 0.9014, 0.6024, 0.448, 0.3616, 0.2564, 0.0862, 0.0062, 0.0038, 0.0006, 0.0002, 0.0, 0.0, 0.0], 4: [0.9898, 0.9668, 0.9068, 0.8852, 0.849, 0.7704, 0.546, 0.3638, 0.318, 0.0866, 0.016, 0.0038, 0.0014, 0.0002, 0.0, 0.0, 0.0], 5: [0.9952, 0.9688, 0.9598, 0.9166, 0.8304, 0.6936, 0.4916, 0.4086, 0.316, 0.0704, 0.013, 0.0046, 0.0014, 0.0, 0.0, 0.0, 0.0], 6: [0.9678, 0.981, 0.927, 0.928, 0.865, 0.7012, 0.5252, 0.4332, 0.2636, 0.092, 0.0126, 0.0022, 0.0006, 0.0, 0.0, 0.0, 0.0], 7: [0.9836, 0.9562, 0.9226, 0.8764, 0.876, 0.726, 0.5374, 0.3614, 0.3218, 0.058, 0.0146, 0.0042, 0.0014, 0.0, 0.0, 0.0, 0.0], 8: [0.9778, 0.9614, 0.9366, 0.9046, 0.8822, 0.7604, 0.5196, 0.3692, 0.2664, 0.0648, 0.0108, 0.003, 0.0014, 0.0, 0.0, 0.0, 0.0], 9: [0.9972, 0.9696, 0.966, 0.908, 0.9056, 0.7056, 0.5192, 0.3878, 0.2856, 0.067, 0.013, 0.0036, 0.0018, 0.0, 0.0, 0.0, 0.0]}
```